src/Microsoft.NetCore.Analyzers/Core/Security/UseAutoValidateAntiforgeryToken.cs

```diff
@@ -340,7 +350,8 @@ void FindAllTheSpecifiedCalleeMethods(ISymbol methodSymbol, HashSet<ISymbol> vis
 
                         foreach (var child in callingMethods.Keys)
                         {
-                            if (onAuthorizationAsyncMethodSymbols.Contains(child))
+                            if (child is IMethodSymbol childMethodSymbol &&
+                                onAuthorizationAsyncMethodSymbols.ContainsKey(childMethodSymbol))
```

Comment:
Do we know where the IndexOutOfRangeException was being thrown? I presume it is in the indexer access below `results[methodSymbol]`? If so, I am not sure how this PR would fix that exception.

Reply:
@LLLXXXCCC based on your guess of the exception is a thread safety issue, I've come up with an unreliable repro (get the rule to trigger when there are multiple controllers in the project).

https://github.com/dotpaul/repro2844

Replace the Microsoft.NetCore.Analyzers.dll with your own build (maybe with Debug.Break() somewhere so you can catch the exceptions), and try building / rebuilding multiple times.

Reply:
Also, don't remember if I got IndexOutOfRangeExceptions, but definitely ran into some others like:

warning AD0001: Analyzer 'Microsoft.NetCore.Analyzers.Security.UseAutoValidateAntiforgeryToken' threw an exception of type 'System.ArgumentException' with message 'Destination array was not long enough. Check destIndex and length, and the array's lower bounds.'.

warning AD0001: Analyzer 'Microsoft.NetCore.Analyzers.Security.UseAutoValidateAntiforgeryToken' threw an exception of type 'System.NullReferenceException' with message 'Object reference not set to an instance of an object.'.

Reply:
I think these might all be likely due to use of non-concurrent data structure in this analyzer.

Reply:
@LLLXXXCCC were you able to repro some exceptions being thrown, and verify your changes don't have exceptions?

Would probably be a good idea to see if a unit test with multiple controllers can repro (and should have one anyway).

Reply:
Yes, I can reproduce the two warnings you mentioned and the IndexOutOfRangeException. For now, the fix works well.

Reply:
Please add a unit test that contains multiple controllers and produces diagnostics, so that we have a test committed that reproduces the original problem.